azure - metric filter - support for child resources

jomalsan opened the pull request:
The following is a proposal for how we can extend the current MetricFilter to make it easy to support child resources. It also includes an example of a metric filter for cosmosdb-databases. This can be tested with the following yaml policy:
```yaml
policies:
  - name: cosmosdb-database-metric
    resource: azure.cosmosdb-database
    filters:
      - type: metric
        metric: TotalRequestUnits
        op: le
        aggregation: total
        threshold: 1000
        timeframe: 24
```

The core of the proposed change is refactoring the portion of `MetricFilter.get_metric_data` that changes between child and parent resources into a function. This way child resources can implement their own metric filter by inheriting from MetricFilter and overriding the method (get_metric_data) needed to implement their custom logic. An example of this is included for CosmosDb.Database

Please comment and ask further questions. I have raw notes going over the different solutions to this problem that I am happy to share, but the proposed solution here seems to be the best. Once everyone has agreed on an implementation I will extend the methodology for other child resources, include tests and new example documentation.

logachev commented on tools/c7n_azure/c7n_azure/filters.py:
```diff
@@ -212,6 +205,16 @@ def get_metric_data(self, resource):
 
         return m
+    
+    def get_metric_data(self, resource):
```
a little confusing what is the difference between `self.filter` and `self.get_filter()`.. 
Maybe create a subclass for all Child resource metrics and use `self.parent_filter` instead of `self.get_filter() ?

jomalsan replied:
I thought about breaking out `get_filter()` but the override in the child class needs to modify the resource id field that is getting passed in and the filter field. This way we maintain one function that needs to be overridden instead of two.

@logachev I think the confusion might be that in this context `self.filter` is an Azure SDK concept and has nothing to do with the c7n concept of filter. An example for Cosmos DB databases would be `self.filter = "DatabaseName eq 'testdatabase'"` which only returns metrics for the given database

stefangordon replied:
Still feels weird to me to do this thing on each child as we'll have lots of places to update it if the API or our functionality changes:

```
class CosmosDBCollectionMetricFilter(MetricFilter):
    def get_metric_data(self, resource):
        if self.filter is None:
            parent_filter = "CollectionName eq '%s'" % resource['id']
        else:
            parent_filter = "%s and CollectionName eq '%s'" % (self.filter, resource['id'])
        return self.client.metrics.list(
            resource['c7n:parent-id'],
            timespan=self.timespan,
            interval=self.interval,
            metricnames=self.metric,
            aggregation=self.aggregation,
            filter=parent_filter
        )
```

Seems like worst case you could do this instead on each resource with less changes to the base and subclass

```
class CosmosDBCollectionMetricFilter(MetricFilter):
    def __init__(self, data, manager=None):
        super(CosmosDBCollectionMetricFilter, self).__init__(data, manager)
        self.filter = ' and '.join(("CollectionName eq '%s'" % resource['id'], self.filter))
        self.id_override =  resource['c7n:parent-id']
```

Then in the base you only have to change `resource['id']` to `self.id_override or resource['id']`

Or you could put the string modification in the base and just set the one field (all children have the `resource['c7n:parent-id']` so you can build the two fields you need if you set this:

```
class CosmosDBCollectionMetricFilter(MetricFilter):
    def __init__(self, data, manager=None):
        super(CosmosDBCollectionMetricFilter, self).__init__(data, manager)
        self.parent_filter_dimension = 'CollectionName'
```

logachev commented on tools/c7n_azure/c7n_azure/resources/cosmos_db.py:
```diff
@@ -172,6 +172,26 @@ def enumerate_resources(self, parent_resource, type_info, **params):
+        if self.filter is None:
+            parent_filter = "DatabaseName eq '%s'" % resource['id']
+        else:
+            parent_filter = "%s and DatabaseName eq '%s'" % (self.filter, resource['id'])
```
I think we need add parentheses: `(%s) and (DatabaseName eq '%s')`

jomalsan replied:
Parentheses are invalid in the filter syntax: https://docs.microsoft.com/en-us/rest/api/monitor/filter-syntax
It looks like the only "complex" logic that is allowed is the "and" operator (not even "or" works), so they just assume that it is `assertion1 and assertion2 and ...`

jomalsan commented on tools/c7n_azure/c7n_azure/resources/cosmos_db.py:
```diff
@@ -172,6 +172,26 @@ def enumerate_resources(self, parent_resource, type_info, **params):
+    """CosmosDB Database Metric Filter
+    """
+    def __init__(self, data, manager=None):
+        super().__init__(data, manager)
```
Removed

stefangordon commented:
I think just add a little test to get coverage for that new subclass and mark it as `ready for review`.    Either a unit test on that one method with a mock on the `list` call, or just a new policy/cassette on the resource type.

jomalsan commented:
> I think just add a little test to get coverage for that new subclass and mark it as `ready for review`. Either a unit test on that one method with a mock on the `list` call, or just a new policy/cassette on this resource type.

I plan on adding the extended subclass to support cosmos db containers as well. Then tests for both and add the examples for both of them

jomalsan commented:
Added child metric for cosmosdb-collections and added tests for both of the new filters.

The cassettes will have to be re-recorded after PR #4748 is merged
